docs/architecture.md

```diff
@@ -0,0 +1,222 @@
+
+Further down the line, consumer stack can be wired with the UCAN Log Stream and use receipts as the trigger.
+
+## DB Stack
+
+The DB stack relies on Amazon RDS Database to keep the necessary state for the w3filecoin pipeline. Its data model was designed with the aim of being the data structure for each of the processors running within the pipeline, while also enabling the tracking of state of each item in the pipeline and to get a mapping between content CIDs and piece CIDs.
+
+### Schema
+
+```sql
+-- Table describes queue of verified CARs to be stored in filecoin
+-- CAR is considered in queue when there is no piece referencing it
+CREATE TABLE content
+(
+  -- CAR CID
+  link TEXT PRIMARY KEY,
+  -- CAR Size
+  size number NOT NULL,
+  -- Source where the content can be fetched from
+  source JSONB NOT NULL,
+  -- Timestamp
+  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL
+);
+
+-- Table describes pieces derived corresponding to CARs in the content table. Link (commP) is
+-- unique even though cargo reference is not, that is because there may be an error in piece
+-- derivation and in that case another correct piece will reference the same content.
+CREATE TABLE piece
+(
+  -- Piece CID
+  link TEXT NOT NULL PRIMARY KEY,
+  -- Piece size
+  size number NOT NULL,
+  -- Reference to the content of this piece (CAR CID).
+  content TEXT NOT NULL REFERENCES content(link),
+  -- Time when the piece was derived from the content
+  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL
+);
+
+CREATE INDEX piece_content_idx ON piece (content);
+CREATE INDEX piece_inserted_idx ON piece (inserted);
+
+-- Content for which we need to derive piece CIDs. We will have a process that
+-- reads from this queue and writes into `piece` table.
+CREATE VIEW content_queue AS
+  SELECT content.*
+  FROM content
+  LEFT OUTER JOIN piece ON content.link = piece.content
+  WHERE piece.content IS NULL
+  ORDER BY piece.inserted;
+
+-- Table describing pieces to be included into aggregates. If aggregate is NULL then the
+-- piece is queued for the aggregation.
+CREATE TABLE inclusion
+(
+  -- Piece CID. Notice that it is not unique because in case of bad piece
+  -- aggregate will be rejected and good pieces will be written back here
+  -- to be included into new aggregate
+  piece TEXT NOT NULL REFERENCES piece(link),
+  
+  -- Aggregate CID, if NULL the the piece is queued for the aggregation
+  aggregate TEXT REFERENCES aggregate(link) NULL,
+
+  -- Priority in the queue. I think it could be a counter on initial inserts hence
+  --  providing FIFO order. When piece is retried (after aggregate is rejected)
+  -- we could keep original priority hence prioritizing it over new items yet
+  -- keeping original FIFO order among rejects.
+  priority TEXT NOT NULL,
```

Comment:
To do what comment describes we need some auto-incrementing counter field.

Reply:
It is not clear yet how priority would work. It would not be an auto-incrementing counter given it is not a table with just auto increment on insert. Made it a number